File: rust/crates/fantastic-io-bridge/src/authorizer/ingress/password.rs
//! `password` — kernel-GROUP membership by a shared secret (ingress side: CHECK).

use super::super::{constant_time_eq, Action, Decision, IngressRule, DEFAULT_TOKEN_ENV};

/// Authorize an inbound `call` only if its envelope `auth_token` matches this leg's
/// group token, read from an env var (`token_env`, default `FANTASTIC_GROUP_TOKEN`)
/// so the secret never touches the portable `.fantastic` workdir. Fail-closed: an
/// unset/empty env var refuses every inbound `call`. Constant-time compare. The
/// egress mirror (`egress::password`) PRESENTS the same token.
pub struct Password {
    token_env: String,
}

impl Password {
    /// Build from an optional `token_env` (default `FANTASTIC_GROUP_TOKEN`).
    pub fn new(token_env: Option<String>) -> Self {
        Self {
            token_env: token_env.unwrap_or_else(|| DEFAULT_TOKEN_ENV.to_string()),
        }
    }

    fn token(&self) -> Option<String> {
        std::env::var(&self.token_env)
            .ok()
            .filter(|s| !s.is_empty())
    }
}

impl IngressRule for Password {
    fn authorize(&self, action: &Action) -> Decision {
        if action.kind != "call" {
            return Decision::Allow; // watch/unwatch already ignored by the read loop
        }
        let Some(expected) = self.token() else {
            return Decision::Deny(format!("group token unset ({})", self.token_env));
        };
        match action.token {
            Some(p) if constant_time_eq(p.as_bytes(), expected.as_bytes()) => Decision::Allow,
            _ => Decision::Deny("invalid or missing group token".to_string()),
        }
    }
}

File: src/lib/python/bundled_agents/io/io_bridge/src/io_bridge/ingress_rules/password.py
"""`password` — kernel-GROUP membership by a shared secret (ingress side: CHECK)."""

from __future__ import annotations

import hmac
import os
from dataclasses import dataclass

from io_bridge._base import ALLOW, Action, Decision, IngressRule


@dataclass(frozen=True)
class Password(IngressRule):
    """Authorize an inbound `call` only if its envelope `auth_token` matches this
    leg's group token, read from an env var (`token_env`, default
    `FANTASTIC_GROUP_TOKEN`) so the secret never touches the portable `.fantastic`
    workdir. Fail-closed: an unset/empty env var refuses every inbound `call`.
    Constant-time compare. The egress mirror (`egress_rules.password`) PRESENTS the
    same token, so one group token in env makes a leg a full member."""

    token_env: str = "FANTASTIC_GROUP_TOKEN"

    def authorize(self, action: Action) -> Decision:
        # Gate EVERY inbound kind the choke point asks about — `call` (dispatch) AND
        # the data-bearing kinds `emit`/`watch`/`state_subscribe`. The web inbound
        # legs route those through this same gate, so a missing token must refuse
        # them too, else a "locked" leg leaks telemetry to a tokenless client.
        expected = os.environ.get(self.token_env) or None
        if expected is None:
            return Decision(False, f"group token unset ({self.token_env})")
        presented = (
            action.token
        )  # the frame-envelope token (NOT the dispatched payload)
        if isinstance(presented, str) and hmac.compare_digest(presented, expected):
            return ALLOW
        return Decision(False, "invalid or missing group token")
